build: use Go 1.23, require minimum 1.22

calmh opened the pull request:
🥳

calmh committed:
build: use Go 1.23, require minimum 1.22

calmh commented:
Windows reparse point handling has changed, we need to check it out:

```
--- FAIL: TestWalkTraverseDirJunct (0.03s)
=== RUN   TestWalkInfiniteRecursion
    walkfs_test.go:132: Infinite recursion not detected correctly
--- FAIL: TestWalkInfiniteRecursion (0.04s)
```

rasa commented:
I spun up a Windows 10 VM (10.0.19045.2965), and the errors appeared. I'll try a Windows 11 VM now...

rasa commented:
The tests pass on a Windows 11 VM Pro 23H2 10.0.22631.2861.

calmh commented:
Well that's worrying.

rasa committed:
Fix walk test failures in Windows 10/2022

Fixes test failures:
=== RUN   TestWalkTraverseDirJunct
    walkfs_test.go:85: Directory junction was not traversed
--- FAIL: TestWalkTraverseDirJunct (0.03s)
=== RUN   TestWalkInfiniteRecursion
    walkfs_test.go:132: Infinite recursion not detected correctly
--- FAIL: TestWalkInfiniteRecursion (0.04s)
raised at
https://github.com/syncthing/syncthing/pull/9651#issuecomment-2286841656
See #9651 for details.

rasa commented:
I committed a change that fixed the test failures on my Windows 10 VM. I haven't tested the change otherwise.

calmh commented:
I don't like that fix. For one, if the symlink bit isn't set (which is indeed a documented change in Go 1.23), we shouldn't need to the reparse handling, we should already traverse the directory as is. Additionally, this would mean we try to parse reparse points on every Lstat which would likely be a significant performance regression.

bt90 commented:
@rasa can you break it down to a minimal reproducer? IMHO this is a go bug.

rasa commented:
@calmh Yes, I will replace that change with a proper one. 

It appears we can replace all or most of our [readReparseTag](https://github.com/syncthing/syncthing/blob/7517d18fbbf34b42aa3d4b231018d2ee704a159d/lib/fs/basicfs_lstat_windows.go#L23), with using [Stat().ReparseTag](https://github.com/golang/go/blob/6885bad7dd86880be6929c02085e5c7a67ff2887/src/os/types_windows.go#L30) instead, which was introduced in [go1.11](https://github.com/golang/go/blob/41e62b8c49d21659b48a95216e3062032285250f/src/os/types_windows.go#L28), as Reserved0, then renamed to ReparseTag in [go1.21rc1](https://github.com/golang/go/blob/1c1c82432a78b06c8010c7257df58ff11cc05b61/src/os/types_windows.go#L28), but not mentioned in the [release notes](https://tip.golang.org/doc/go1.21#minor_library_changes).

I've assigned this to myself, and will test the code on both my VMs, as well as GitHub's windows-2022 and windows-2019 instances.

@bt90 I guess your right that it's a go bug. Go should return consistent results regardless of what version of Windows its run on.

rasa commented:
A solution is forthcoming. It doesn't involve Win32 API calls so our `readReparseTag` function will go away. Here's a preview:
```go
func (f *BasicFilesystem) underlyingLstat(name string) (os.FileInfo, error) {
	var fi, err = os.Lstat(name)

	// There are cases where files are tagged as symlink, but they end up being
	// something else. Make sure we properly handle those types.
	if err == nil {
		if f.junctionsAsDirs {
			var checkIfJunction bool
			// godebug("winsymlink") will return >0 if GODEBUG=winsymlink=0 is set
			if compare.Compare(runtime.Version(), "go1.23") >= 0 && godebug("winsymlink") == 0 {
				checkIfJunction = fi.Mode()&os.ModeIrregular != 0
			} else {
				checkIfJunction = fi.Mode()&os.ModeSymlink != 0
			} else {

			// NTFS directory junctions can be treated as ordinary directories,
			// see https://forum.syncthing.net/t/option-to-follow-directory-junctions-symbolic-links/14750
			if checkIfJunction {
				if reparseTag, reparseErr := readReparseTag(name); reparseErr == nil && isDirectoryJunction(reparseTag) {
					return &dirJunctFileInfo{fi}, nil
				}
			}
		}

		// Workaround for #9120 till golang properly handles deduplicated files by
		// considering them regular files.
		if fi.Mode()&os.ModeIrregular != 0 {
			if reparseTag, reparseErr := readReparseTag(name); reparseErr == nil && isDeduplicatedFile(reparseTag) {
				return &dedupFileInfo{fi}, nil
			}
		}
	}

	return fi, err
}

func godebug(godebug string) uint64 {
	myMetric := "/godebug/non-default-behavior/" + godebug + ":events"

	// Create a sample for the metric.
	sample := make([]metrics.Sample, 1)
	sample[0].Name = myMetric

	// Sample the metric.
	metrics.Read(sample)

	// Check if the metric is actually supported.
	// If it's not, the resulting value will always have
	// kind KindBad.
	if sample[0].Value.Kind() == metrics.KindBad {
		return 0
	}
	// Handle the result.
	// It's OK to assume a particular Kind for a metric;
	// they're guaranteed not to change.
	return sample[0].Value.Uint64()
}
```

rasa commented:
The below code [works](https://github.com/rasa/syncthing/actions/runs/10437170537) in go1.21, go1.22, and go1.23, on windows 10, 11, windows-latest (windows-2022) and windoes-2019, with or without `GODEBUG=winsymlink=0` set:

```go
func (f *BasicFilesystem) underlyingLstat(name string) (os.FileInfo, error) {
	fi, err := os.Lstat(name)

	if err == nil && f.junctionsAsDirs {
		v := reflect.Indirect(reflect.ValueOf(fi))
		y := v.FieldByName("ReparseTag")
		if y.Interface() == uint32(windows.IO_REPARSE_TAG_MOUNT_POINT) {
			return &dirJunctFileInfo{fi}, nil
		}
	}

	return fi, err
}
```
Is it acceptable to use reflect to access non-public members of a FileInfo? It sure makes things faster and simpler.

rasa commented:
Another solution that would work for at least [two years](https://tip.golang.org/doc/godebug) would be to add
```go
//go:debug winsymlink=0
```
to the existing file. This would force the pre-go1.23 functionality.

calmh commented:
I would prefer we understand what's going on and do the minimal fix. We shouldn't need to change our reparse parsing stuff, only when to trigger it. In principle it seems to me this should be just adding a "if go1.22" to our existing symlink-to-reparse-point check and in go1.23+ do nothing at all as it should just look like a normal directory.

rasa committed:
Revert 3bca8ab9 per @calmh's comment

See https://github.com/syncthing/syncthing/pull/9651#issuecomment-2294709518

rasa committed:
Fix walk test failures in Windows 10/2022

This fixes the test failures:
=== RUN   TestWalkTraverseDirJunct
    walkfs_test.go:85: Directory junction was not traversed
--- FAIL: TestWalkTraverseDirJunct (0.03s)
=== RUN   TestWalkInfiniteRecursion
    walkfs_test.go:132: Infinite recursion not detected correctly
--- FAIL: TestWalkInfiniteRecursion (0.04s)
documented in #9651.

This also undoes #9168 by not checking if ModeIrregular files are dedups as
https://github.com/golang/go/commit/f26fa055229502fb6df4b6d9c300eccf01098248
removed the ModeIrregular bit from dedup files.

calmh commented on lib/fs/basicfs_lstat_windows.go:
```diff
@@ -96,19 +93,11 @@ func (f *BasicFilesystem) underlyingLstat(name string) (os.FileInfo, error) {
 			}
 		}
-
-		// Workaround for #9120 till golang properly handles deduplicated files by
-		// considering them regular files.
-		if fi.Mode()&os.ModeIrregular != 0 {
-			if reparseTag, reparseErr := readReparseTag(name); reparseErr == nil && isDeduplicatedFile(reparseTag) {
-				return &dedupFileInfo{fi}, nil
-			}
-		}
```
Was this fixed in 1.22 and earlier? I did a search but I don't find the fix for this documented

rasa replied:
It was fixed in 1.21 in Nov 2023 in
https://github.com/golang/go/commit/f26fa055229502fb6df4b6d9c300eccf01098248

calmh commented on lib/fs/basicfs_lstat_windows.go:
```diff
@@ -60,32 +60,29 @@ func isDirectoryJunction(reparseTag uint32) bool {
-	// irregular file.
-	return fi.FileInfo.Mode() &^ os.ModeIrregular
+func init() {
+	junctionPointModeMask = os.ModeSymlink
+	if version.Compare(runtime.Version(), "go1.23") >= 0 {
+		// if GODEBUG=winsymlink=0, then we are in g1.22 mode so let's check for
+		// os.ModeSymlink as well, as it can't hurt.
+		junctionPointModeMask |= os.ModeIrregular
+	}
```
👍
